tmva/pymva/src/MethodPyKeras.cxx

```diff
@@ -154,51 +155,132 @@ UInt_t TMVA::MethodPyKeras::GetNumValidationSamples()
+   // Setup model, either the initial model from `fFilenameModel` or
+   // the trained model from `fFilenameTrainedModel`
+   if (fContinueTraining) Log() << kINFO << "Continue training with trained model" << Endl;
+   SetupKerasModel(fContinueTraining);
+}
 
-   // check first if using tensorflow backend
-   if (GetKerasBackend() == kTensorFlow) {
-      Log() << kINFO << "Using TensorFlow backend - setting special configuration options "  << Endl;
-      PyRunString("import tensorflow as tf","Error importing tensorflow");
-      PyRunString("from keras.backend import tensorflow_backend as K");
-      // run these above lines also in global namespace to make them visible overall
-      PyRun_String("import tensorflow as tf", Py_single_input, fGlobalNS, fGlobalNS);
-      PyRun_String("from keras.backend import tensorflow_backend as K", Py_single_input, fGlobalNS, fGlobalNS);
+void MethodPyKeras::SetupKerasModel(bool loadTrainedModel) {
+
+   // initialize first Keras. This is done only here when class has
+   // all state variable set from options or read from XML file
+   // Import Keras
+
+   bool useTFBackend = kFALSE;
+   bool kerasIsCompatible = kTRUE;
+   bool kerasIsPresent = kFALSE;
+
+   if (!fUseTFKeras) {
+      auto ret  = PyRun_String("import keras", Py_single_input, fGlobalNS, fLocalNS);
+      // need importing also in global namespace
+      if (ret != nullptr) ret = PyRun_String("import keras", Py_single_input, fGlobalNS, fGlobalNS);
+      if (ret != nullptr)
+         kerasIsPresent = kTRUE;
+      if (kerasIsPresent) {
+         // check compatibility with tensorflow
+         if (GetKerasBackend() == kTensorFlow ) {
+            useTFBackend = kTRUE;
+
+            PyRunString("keras_major_version = int(keras.__version__.split('.')[0])");
+            PyRunString("keras_minor_version = int(keras.__version__.split('.')[1])");
+            PyObject *pyKerasMajorVersion = PyDict_GetItemString(fLocalNS, "keras_major_version");
+            PyObject *pyKerasMinorVersion = PyDict_GetItemString(fLocalNS, "keras_minor_version");
+            int kerasMajorVersion = PyLong_AsLong(pyKerasMajorVersion);
+            int kerasMinorVersion = PyLong_AsLong(pyKerasMinorVersion);
+            Log() << kINFO << "Using Keras version " << kerasMajorVersion << "." << kerasMinorVersion << Endl;
+            kerasIsCompatible = (kerasMajorVersion >= 2 && kerasMinorVersion >= 3);
 
+         }
+      } else {
+         // Keras is not found. try tyo use tf.keras
+         Log() << kINFO << "Keras is not found. Trying using tf.keras" << Endl;
+         fUseTFKeras = 1;
+      }
+   }
+
+   // import Tensoprflow (if requested or because is keras backend)
+   if (fUseTFKeras || useTFBackend) {
+      auto ret = PyRun_String("import tensorflow as tf", Py_single_input, fGlobalNS, fLocalNS);
+      if (ret != nullptr) ret = PyRun_String("import tensorflow as tf", Py_single_input, fGlobalNS, fGlobalNS);
+      if (ret == nullptr) {
```

Comment:
This can be an `else` clause.

Reply:
I don;t see the problem here. `ret` is updated in line 218 and it would not be done if I add an `else` clause

Reply:
Yep, right!